tests/framework/crossplane.go

```diff
@@ -39,40 +41,74 @@ type ExpectedNginxField struct {
 // ValidateNginxFieldExists accepts the nginx config and the configuration for the expected field,
 // and returns whether or not that field exists where it should.
 func ValidateNginxFieldExists(conf *Payload, expFieldCfg ExpectedNginxField) error {
+	var directiveFoundInServer, directiveFoundInUpstream bool
+
+	b, err := json.Marshal(conf)
+	if err != nil {
+		return fmt.Errorf("error marshaling nginx config: %w", err)
+	}
+
 	for _, config := range conf.Config {
 		if !strings.Contains(config.File, expFieldCfg.File) {
 			continue
 		}
 
 		for _, directive := range config.Parsed {
-			if len(expFieldCfg.Servers) == 0 {
+			if len(expFieldCfg.Server) == 0 && len(expFieldCfg.Upstream) == 0 {
 				if expFieldCfg.fieldFound(directive) {
 					return nil
 				}
 				continue
 			}
 
-			for _, serverName := range expFieldCfg.Servers {
-				if directive.Directive == "server" && getServerName(directive.Block) == serverName {
-					for _, serverDirective := range directive.Block {
-						if expFieldCfg.Location == "" && expFieldCfg.fieldFound(serverDirective) {
-							return nil
-						} else if serverDirective.Directive == "location" &&
-							fieldExistsInLocation(serverDirective, expFieldCfg) {
-							return nil
-						}
-					}
-				}
+			directiveFoundInServer = validateServerBlockDirectives(expFieldCfg, *directive)
+
+			directiveFoundInUpstream = validateUpstreamDirectives(expFieldCfg, *directive)
+
+			if len(expFieldCfg.Server) > 0 && directiveFoundInServer {
+				return nil
+			}
+
+			if len(expFieldCfg.Upstream) > 0 && directiveFoundInUpstream {
+				return nil
 			}
 		}
 	}
 
-	b, err := json.Marshal(conf)
-	if err != nil {
-		return fmt.Errorf("error marshaling nginx config: %w", err)
+	return fmt.Errorf("directive %s not found in: nginx config %s", expFieldCfg.Directive, string(b))
+}
+
+func validateServerBlockDirectives(
+	expFieldCfg ExpectedNginxField,
+	directive Directive,
+) bool {
+	var fieldFound bool
+	if directive.Directive == "server" && getServerName(directive.Block) == expFieldCfg.Server {
+		for _, serverDirective := range directive.Block {
+			if expFieldCfg.Location == "" && expFieldCfg.fieldFound(serverDirective) {
+				fieldFound = true
+			} else if serverDirective.Directive == "location" &&
+				fieldExistsInLocation(serverDirective, expFieldCfg) {
+				fieldFound = true
+			}
+		}
 	}
+	return fieldFound
+}
 
-	return fmt.Errorf("field not found; expected: %+v\nNGINX conf: %s", expFieldCfg, string(b))
+func validateUpstreamDirectives(
+	expFieldCfg ExpectedNginxField,
+	directive Directive,
+) bool {
+	var fieldFound bool
+	if directive.Directive == "upstream" && directive.Args[0] == expFieldCfg.Upstream {
+		for _, directive := range directive.Block {
+			if expFieldCfg.fieldFound(directive) {
+				fieldFound = true
```

Comment:
We can just return true here, then we don't need the `fieldFound` variable.

Reply:
Yes initially was doing that. So when iterating over a block, if the first directive doesn't match, it returns prematurely. Added these fields so it goes through the entire block, before declaring if it found the directive or not

Reply:
Oh yes, you're right. I totally misread it haha.

Reply:
Oh wait no, I think my comment still stands. We're only returning true if we found the field, otherwise we don't hit this conditional and we go through another iteration.

Reply:
oh yeah sorry you are right ! maybe I was testing with other things.